### DISCUSSION TITLE
How to rebuild the _ext.cpp when developing?

### DISCUSSION THREAD

**Discussion by kawaho** [scikit-hep/fastjet#302] [Q&A] [answered]
Sorry this is a really basic question. I am trying to add some features into _ext.cpp. How do I rebuild the package after modifying the file?

In the README.md, it says

```bash python -m pip install '.[test]'```

but I assume this is just for the first time building the package. I imagine there is a way to only recompile the changes I did in the .cpp files?

**Comment by matthewfeickert** [comment]
> I am trying to add some features into _ext.cpp.

@kawaho Please open up an Issue (https://github.com/scikit-hep/fastjet/issues) that describes the features that you want to add, so these can get discussed in advance.

> How do I rebuild the package after modifying the file?

Building happens during the install, so just install your modified version with

```
python -m pip install --upgrade .
```

**Comment by kawaho** [comment]
Hi @matthewfeickert , coming back to this question, I was not able to rebuild the package using  

```python -m pip install --upgrade .```  

When I do so, I get the following error:

``` running build_ext
  building 'fastjet._ext' extension
  /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/bin/x86_64-conda-linux-gnu-c++ -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/include -fPIC -I/isilon/export/home/kho29/fastjet_dev/fastjet/src/fastjet/_fastjet_core/include -I/tmp/pip-build-env-1urvtoly/overlay/lib/python3.12/site-packages/pybind11/include -I/isilon/export/home/kho29/mambaforge/envs/fastjet_dev/include/python3.12 -c src/_ext.cpp -o build/temp.linux-x86_64-cpython-312/src/_ext.o -fvisibility=hidden -g0 -std=c++11
  /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/bin/x86_64-conda-linux-gnu-c++ -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/include -pthread -B /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/compiler_compat -shared -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/isilon/export/home/kho29/mambaforge/envs/fastjet_dev/lib -Wl,-rpath-link,/isilon/export/home/kho29/mambaforge/envs/fastjet_dev/lib -L/isilon/export/home/kho29/mambaforge/envs/fastjet_dev/lib -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/include build/temp.linux-x86_64-cpython-312/src/_ext.o -L/isilon/export/home/kho29/fastjet_dev/fastjet/src/fastjet/_fastjet_core/lib -Wl,-R$ORIGIN/_fastjet_core/lib -lfastjet -lfastjettools -lsiscone -lsiscone_spherical -lfastjetplugins -lfastjetcontribfragile -o build/lib.linux-x86_64-cpython-312/fastjet/_ext.cpython-312-x86_64-linux-gnu.so
  installing to build/bdist.linux-x86_64/wheel
  running install
  error: [Errno 17] File exists: 'build/lib.linux-x86_64-cpython-312/fastjet/_fastjet_core'
  error: subprocess-exited-with-error
  
  × Building wheel for fastjet (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/bin/python /isilon/export/home/kho29/mambaforge/envs/fastjet_dev/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py build_wheel /tmp/tmpgrzugkax
  cwd: /isilon/export/home/kho29/fastjet_dev/fastjet
  Building wheel for fastjet (pyproject.toml) ... error
  ERROR: Failed building wheel for fastjet
Failed to build fastjet
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (fastjet)
```

If I manually remove ```build``` and ```src/fastjet/_fastjet_core/```, everything gets built from scratch instead of just updating my modifications. Any suggestions on this? Thanks!

**Reply by matthewfeickert** [reply]
> If I manually remove build and src/fastjet/_fastjet_core/, everything gets built from scratch instead of just updating my modifications.

@kawaho Are you able to build and install before your modifications or is everything broken as of now? If everything is broken, please open an Issue with a minimal failing environment description.

If things are only happening post your changes please provide some context around the state of your repository, as this error message in the abstract isn't very informative.

**Reply by kawaho** [reply]
To clarify, I tried two things:

1. build -> make modifications -> do ```python -m pip install --upgrade .``` -> got the error above
2. build -> make modifications -> remove ```build```  and ```src/fastjet/_fastjet_core/``` -> build from scratch with ```python -m pip install '.[test]'``` -> works

But I don't really want the second way as I only want to update my modified part.

**Reply by matthewfeickert** [reply]
@kawaho You might be hitting Issue https://github.com/scikit-hep/fastjet/issues/279. Try something similar there, but note that depending on where you're making changes you'll want to be sure you're committing your work.

**Reply by kawaho** [reply]
@matthewfeickert thanks for the suggestions. On top of the script you provided, I also needed to do ```rm -rf src/fastjet/_fastjet_core/``` to build successfully.

**Reply by matthewfeickert** [reply]
> On top of the script you provided, I also needed to do `rm -rf src/fastjet/_fastjet_core/` to build successfully.

You should not be removing the FastJet C++ source code. This is definitely not correct.

Please explain what it is you're actually trying to do and what the state of the repository is in. Though again, if you want to contribute additions to the source code open an Issue first so that there can be a discussion in advance.

**edit:** (Rereading the whole issue carefully this time to pay attention to all your questions &mdash; sorry)

> I imagine there is a way to only recompile the changes I did in the .cpp files?
... 
> everything gets built from scratch instead of just updating my modifications

Ah, no. You are probably already aware of this, so apologies if I'm telling you things you already know, but, quoting @jpivarski, incremental compilation isn't something that `pip install` likes to do as it the opposite of encapsulation.

Further quoting Jim,

> To actually do incremental compilation, the pure-FastJet part would have to be separated from the `pybind11` part that is built by asking `pip` to build from a pre-built library, which would only make the build procedure more complicated.

and so is something that would really require Issue #310 to get dealt with.

**Reply by kawaho** [reply]
Hi @matthewfeickert ,

> You should not be removing the FastJet C++ source code. This is definitely not correct.

This is the part where I got confused because there are ```fastjet-core/``` and ```src/fastjet/_fastjet_core```. The latter is created after the build so I thought it's ok to remove it as ```pip``` complains something like  ```src/fastjet/_fastjet_core already exists``` .

> Please explain what it is you're actually trying to do and what the state of the repository is in. Though again, if you want to contribute additions to the source code open an Issue first so that there can be a discussion in advance.

Yes, I have opened an issue some time ago: https://github.com/scikit-hep/fastjet/issues/303

**Reply by matthewfeickert** [reply]
> The latter is created after the build so I thought it's ok to remove it as `pip` complains something like `src/fastjet/_fastjet_core already exists` .

@kawaho My apologies, you're absolutely right, and even though you pointed everything out clearly and I said I carefully reread things I still misread the paths. Sorry.

You're correct that do do rebuilds you would need to remove this transient directory. This, again, is not something that you as a developer should be responsible for manually having to keep track of, so we again see the need for Issue #279 and Issue #310 to get dealt with.

**Reply by matthewfeickert** [reply]
I've opened PR https://github.com/scikit-hep/fastjet/pull/311 to partially address parts of this.

**Reply by matthewfeickert** [reply, marked as answer]
Between PR #311 and https://github.com/scikit-hep/fastjet/issues/279#issuecomment-2347977214 I think this is now fully resolved so that one can now just run

```
python -m pip install --upgrade .
```

from the top level directory and get a working rebuild. Until Issue #279 is fully resolved, to get a totally clean slate for all of the Git submodules you can run https://github.com/scikit-hep/fastjet/issues/279#issuecomment-2347977214 in advance of install.
